### DISCUSSION TITLE
js simplify with option for ranges would have been nice

### DISCUSSION THREAD

**Discussion by vegarringdal** [zeux/meshoptimizer#716] [Ideas]
Thanks for making this amazing library and sharing it. 🥇 

Maybe my idea is dumb, but Ill ask anyways :-)

In my app I have merged meshes, and I know the draw ranges of each item.
I use this with a custom batchedmesh in threejs, batchedmesh is using ([WEBGL_multi_draw extension](https://developer.mozilla.org/en-US/docs/Web/API/WEBGL_multi_draw))

In meshoptimizer we atm have this:

```ts
function simplify: (
   indices: Uint32Array, 
   vertex_positions: Float32Array, 
   vertex_positions_stride: number, 
   target_index_count: number, 
   target_error: number, flags?: Flags[] | undefined
) => [newIndex, error]
```

Would have been nice to be able to send in ranges of items, and it could use this information with "lockedborder" flag

```ts
function simplifyX: (
   indices: Uint32Array, 
   vertex_positions: Float32Array, 
   firsts: Uint32Array,  // new...
   counts : Uint32Array,  // new..
   vertex_positions_stride: number, 
   target_index_count: number, 
   target_error: number, flags?: Flags[] | undefined
) => [new_indicies, new_start, new_count, error]
```

Or maybe it could aslo return back cleaned up position array too ?

```ts
function simplifyY: (
   indices: Uint32Array, 
   vertex_positions: Float32Array, 
   firsts: Uint32Array,  // new...
   counts : Uint32Array,  // new..
   vertex_positions_stride: number, 
   target_index_count: number, 
   target_error: number, flags?: Flags[] | undefined
) => [new_indicies, new_positions, new_start, new_count, error]
```

Maybe option we could pass in to tell its allowed to merge position of borders if they are closer than x distance

```ts
function simplifyZ: (
   indices: Uint32Array, 
   vertex_positions: Float32Array, 
   firsts: Uint32Array,  // new...
   counts : Uint32Array,  // new..
   vertex_positions_stride: number, 
   target_index_count: number, 
   merge_vertex_position: number, // new
   target_error: number, flags?: Flags[] | undefined
) => [new_indicies, new_positions, new_start, new_count, error]
```

**Comment by zeux** [comment]
When processing a range of indices (eg used for multidraw), you should be able to simplify each subset individually by calling simplify on a slice of the index array. You might want to also use the recently added "Sparse" flag when doing this (add it to flags array) to accelerate processing. If the ranges have connected elements (eg if they represent material splits), "LockBorder" is also good to use in these situations.
